src/app/qgsmaptoolsplitfeatures.cpp

```diff
@@ -113,7 +113,14 @@ void QgsMapToolSplitFeatures::cadCanvasReleaseEvent( QgsMapMouseEvent *e )
     const bool topologicalEditing = QgsProject::instance()->topologicalEditing();
     QgsPointSequence topologyTestPoints;
     vlayer->beginEditCommand( tr( "Features split" ) );
-    const Qgis::GeometryOperationResult returnCode = vlayer->splitFeatures( captureCurve(), topologyTestPoints, true, topologicalEditing );
+
+    // we need to drop Z value to properly split 3D feature in 2D. If not, generated vertices Z value
+    // will be assigned with the mean value between the interpolated Z values of the intersecting point
+    // and the default Z value, which is not what we want
+    QgsCompoundCurve *curve = captureCurve()->clone();
```

Comment:
There's a leak here:
```suggestion
    std::unique_ptr< QgsCompoundCurve > curve( captureCurve()->clone() );
```